edge-util/src/Microsoft.Azure.Devices.Edge.Util/EventHubPartitionKeyResolver.cs

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Util
+{
+    using System;
+    using System.Globalization;
+    using System.Text;
+
+    public static class EventHubPartitionKeyResolver
+    {
+        const short DefaultLogicalPartitionCount = short.MaxValue;
+
+        public static string ResolveToPartition(string partitionKey, int partitionCount)
+        {
+            if (string.IsNullOrWhiteSpace(partitionKey))
+            {
+                throw new ArgumentNullException("partitionKey");
+            }
+
+            if (partitionCount < 1 || partitionCount > DefaultLogicalPartitionCount)
+            {
+                throw new ArgumentOutOfRangeException("partitionCount", partitionCount, string.Format(CultureInfo.InvariantCulture, "Should be between {0} and {1}", 1, DefaultLogicalPartitionCount));
+            }
+
+            short logicalPartition = Math.Abs((short)(PerfectHash.HashToShort(partitionKey) % DefaultLogicalPartitionCount));
+
+            int shortRangeWidth = (int)Math.Floor((decimal)DefaultLogicalPartitionCount / (decimal)partitionCount);
+            int remainingLogicalPartitions = DefaultLogicalPartitionCount - (partitionCount * shortRangeWidth);
+            int largeRangeWidth = shortRangeWidth + 1;
+            int largeRangesLogicalPartitions = largeRangeWidth * remainingLogicalPartitions;
+            int partitionIndex = logicalPartition < largeRangesLogicalPartitions
+                ? logicalPartition / largeRangeWidth
+                : remainingLogicalPartitions + ((logicalPartition - largeRangesLogicalPartitions) / shortRangeWidth);
+
+            return partitionIndex.ToString(NumberFormatInfo.InvariantInfo);
+        }
+    }
+
+    public static class PerfectHash
```

Comment:
class comments would be nice, like why do we need custom hash?

Reply:
also, is this our invention or we took it from somewhere? do we have tests for that?

Reply:
nit: I wonder why they could not keep them public for us and why they made such breaking change :)